File: java/src/main/java/it/giannibombelli/wsc2026/payment/application/commands/AcceptTransaction.java
package it.giannibombelli.wsc2026.payment.application.commands;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.Command;
import it.giannibombelli.wsc2026.common.domain.primitive.Timestamp;
import it.giannibombelli.wsc2026.payment.domain.payment.PaymentId;
import it.giannibombelli.wsc2026.payment.domain.payment.TransactionId;

public record AcceptTransaction(PaymentId aggregateId, TransactionId transactionId, Timestamp providerCompletedAt)
    implements Command<PaymentId> {
    public AcceptTransaction {
        Require.requireArgument(aggregateId, "aggregateId");
        Require.requireArgument(transactionId, "transactionId");
        Require.requireArgument(providerCompletedAt, "providerCompletedAt");
    }
}

File: typescript/src/payment/application/commands/acceptTransaction.ts
// Command per accettare una transazione.

import type { Command } from '@/common/application/command.js';
import type { Timestamp } from '@/common/domain/primitive/timestamp.js';
import type { PaymentId } from '../../domain/payment/paymentId.js';
import type { TransactionId } from '../../domain/payment/transactionId.js';
import { requireArgument } from '@/common/utils/requireArgument.js';

export type AcceptTransaction = Command<'AcceptTransaction', PaymentId> & {
  readonly transactionId: TransactionId;
  readonly providerCompletedAt: Timestamp;
};

export function acceptTransaction(
  aggregateId: PaymentId,
  transactionId: TransactionId,
  providerCompletedAt: Timestamp
): AcceptTransaction {
  requireArgument(aggregateId, 'aggregateId');
  requireArgument(transactionId, 'transactionId');
  requireArgument(providerCompletedAt, 'providerCompletedAt');
  return { kind: 'AcceptTransaction', aggregateId, transactionId, providerCompletedAt };
}
